### PR TITLE
OSDOCS-3544 & 3545: CSI Migration of Cinder and Azure Disk to GA

### PR DESCRIPTION
**Pull request by lpettyjo**
Version(s): 4.11+

Issue:[OSDOCS-3544 ](https://issues.redhat.com/browse/OSDOCS-3544)O[SDOCS-3545](https://issues.redhat.com/browse/OSDOCS-3545)

Link to docs preview: http://file.rdu.redhat.com/lpettyjo/OSDOCS-3544and3545/storage/container_storage_interface/persistent-storage-csi-migration.html

Additional information: PTAL @bertinatto @dobsonj @ropatil010 @gcharot

### CODE REVIEW

File: modules/persistent-storage-csi-migration-overview-support-level.adoc [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+//
+// * storage/container_storage_interface/persistent-storage-csi-migration.adoc
+
+:_content-type: CONCEPT
+[id="persistent-storage-csi-migration-overview-support-level_{context}"]
+= Support level
+
+Certain in-tree volume plug-ins and their equivalent Container Storage Interface (CSI) driver are supported in Technology Preview (TP) status, whereas others are supported in General Availability (GA) status.
+
+:FeatureName: CSI automatic migration for some in-tree volume plug-ins/CSI driver pairs
```
Comment (kalexand-rh) [inline modules/persistent-storage-csi-migration-overview-support-level.adoc:11]: Per the ISG:
> In most cases, do not use slashes in text because they can be ambiguous.

File: modules/persistent-storage-csi-migration-overview.adoc [hunk 2]
```diff
@@ -6,27 +6,6 @@
 [id="persistent-storage-csi-migration-overview_{context}"]
 = Overview
 
-With the automatic migration feature enabled, volumes provisioned using in-tree storage plug-ins supported by this feature are migrated to their counterpart CSI drivers.
+Volumes provisioned using in-tree storage plug-ins supported by this feature are migrated to their counterpart Container Storage Interface (CSI) drivers. This process does not involve any data migration, {product-title} only translates the persistent volume object in memory. As a result, the translated persistent volume object is not stored on disk, nor is its contents changed.
```
Comment (kalexand-rh) [inline modules/persistent-storage-csi-migration-overview.adoc:9]: Per the ISG:
> ! using noun
To avoid ambiguity, replace this gerund with either "by using" or "that uses".

I also suggest simplifying the last sentence.

```suggestion
Volumes provisioned by using in-tree storage plug-ins supported by this feature are migrated to their counterpart Container Storage Interface (CSI) drivers. This process does not involve any data migration. {product-title} only translates the persistent volume object in memory. As a result, the translated persistent volume object is not stored on disk, and its contents are not changed.
```

File: modules/persistent-storage-csi-migration-overview.adoc [hunk 3]
```diff
@@ -6,27 +6,6 @@
...
-
-CSI automatic migration should be seamless. Enabling this feature does not change how you use all existing API objects (for example, `PersistentVolumes`, `PersistentVolumeClaims`, and `StorageClasses`).
-
-By default, automatic migration is disabled.
-
-[IMPORTANT]
-====
-CSI automatic migration will be enabled by default in a future {product-title} release, so it is highly recommended that you test it now and report any issues.
-====
+CSI automatic migration should be seamless. This feature does not change how you use all existing API objects (for example, `PersistentVolumes`, `PersistentVolumeClaims`, and `StorageClasses`).
```
Comment (kalexand-rh) [inline modules/persistent-storage-csi-migration-overview.adoc:11]: Per the ISG:

> Avoid using parentheses in general text because text in parentheses is decreased in importance.

> ! should verb
Whenever possible, rewrite in a more direct way. Do not use "should" to refer to an action that must be performed.

**Review by kalexand-rh — COMMENTED** [review]
I see a few small ISG issues and have a couple of simplification suggestions. This is looking good, though!

File: modules/persistent-storage-csi-migration-overview-support-level.adoc [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+include::snippets/technology-preview.adoc[leveloffset=+1]
+
+The following table provides details about the support level of in-tree volume plug-ins/CSI driver pairs.
+
+.CSI automatic migration In-tree volume plug-ins/CSI driver pair support in {product-title}
+[cols=",^v,^v,^v width="100%",options="header"]
+|===
+|CSI drivers |Support level |CSI Auto migration enabled automatically?
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-overview-support-level.adoc:19]: I would change "support level" to "Migration support level" in to order to avoid potential confusion on the CSI driver itself being TP.

File: modules/persistent-storage-csi-migration-enable.adoc [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 [NOTE]
 ====
-Enabling CSI automatic migration drains, and then restarts, all nodes in the cluster in sequence. This might take some time.
+Enabling Container Storage Interface (CSI) automatic migration drains, and then restarts, all nodes in the cluster in sequence. This might take some time.
```
Comment (gcharot) [inline modules/persistent-storage-csi-migration-enable.adoc:11]: I would highlight the fact that this is section is only intended for testing purposes (testing CSI migration before being GA). For production system, the CSI migration will be automatically enabled and no manual operation is required (once the migration is GA). It's mentioned earlier in the doc but it would be best to avoid any confusion.

File: modules/persistent-storage-csi-migration-overview.adoc (hunk 2 above)
Comment (bertinatto) [inline modules/persistent-storage-csi-migration-overview.adoc:9]: @lpettyjo could you add a section explaining how the default StorageClass is determined? Basically:

During TechPreview, the in-tree StorageClass is the default one. For GA:
1. The CSI SC will be the default in new 4.11 clusters.
2. The in-tree SC will remain the default one if the cluster was upgraded from 4.10. However, the CSI driver will be used under the hood (as opposed to the in-tree plugin).

File: modules/persistent-storage-csi-migration-overview.adoc (hunk 3 above)
Comment (gnufied) [inline modules/persistent-storage-csi-migration-overview.adoc:11]: Should we highlight the fact that - enabling CSI migration for intree PV or PVCs does not enable new CSI driver features? Such as - existing intree PV/PVCs can not be snapshotted or expanded (if intree volume did not support it).

**Comment by gnufied** [comment]
Another thing worth noting is - the final document page has too many "Important !" exclamation marks which IMO hinder readability of final page.
Screenshot from 2022-04-29 10-37-01

File: modules/persistent-storage-csi-migration-enable.adoc [hunk 6]
```diff
@@ -20,7 +20,7 @@ Enabling CSI automatic migration drains, and then restarts, all nodes in the clu
 After turning on Technology Preview features using feature gates, they cannot be turned off. As a result, cluster upgrades are prevented.
 ====
 +
-The following configuration example enables CSI automatic migration to all CSI drivers supported by this feature:
+The following configuration example enables CSI automatic migration for all CSI drivers supported by this feature that are currently in Technology Preview (TP) status:
 +
```
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-enable.adoc:24]: I've added a paragraph at the beginning of the Enabling section to highlight this.

File: modules/persistent-storage-csi-migration-overview.adoc (hunk 3 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-overview.adoc:11]: I've added this information to the Overview section.

File: modules/persistent-storage-csi-migration-enable.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-enable.adoc:26]: This is now covered in the extensive re-write that I've done.

**Comment by lpettyjo** [comment]
The important note issue has been resolved as well though the re-write.

File: storage/container_storage_interface/persistent-storage-csi-migration.adoc [hunk 7]
```diff
@@ -6,13 +6,17 @@ include::_attributes/common-attributes.adoc[]
...
-include::snippets/technology-preview.adoc[leveloffset=+1]
+In-tree storage drivers that are traditionally shipped with {product-title} are being deprecated and replaced by their equivalent Container Storage Interface (CSI) drivers. {product-title} provides automatic migration for certain supported in-tree volume plug-ins to their equivalent CSI drivers.
 
 include::modules/persistent-storage-csi-migration-overview.adoc[leveloffset=+1]
 
+=== Additional resources
+* xref:../../storage/container_storage_interface/persistent-storage-csi-migration.adoc#persistent-storage-csi-migration-automatic-ga[Automatic migration of in-tree volumes to CSI]
+
+* xref:../../storage/container_storage_interface/persistent-storage-csi-migration.adoc#persistent-storage-csi-migration-enable_persistent-storage-csi-migration[Manually enabling CSI automatic migration]
+
```
Comment (gnufied) [inline storage/container_storage_interface/persistent-storage-csi-migration.adoc:17]: The additional resources pages refers to itself, which is kinda weird.

File: modules/persistent-storage-csi-migration-automatic-ga.adoc [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+{product-title} supports automatic and seamless migration for the following in-tree volume types to their Container Storage Interface (CSI) driver counterpart:
+
+* Azure Disk
+
+* OpenStack Cinder
+
+CSI migration for these volume types is considered Generally Available (GA), and requires no manual intervention.
+
+Upgrading from {product-title} versions 4.10, and earlier, to 4.11, or later, automatically migrates existing in-tree volumes so that they are managed by their equivalent CSI driver.
```
Comment (gnufied) [inline modules/persistent-storage-csi-migration-automatic-ga.adoc:17]: We have missed some details about storageclasses from https://github.com/openshift/openshift-docs/pull/45281/files#diff-1116fd8ef55d77a7e4858d07631693b56bce7d56792872528a63173f7b3432b7R25

File: storage/container_storage_interface/persistent-storage-csi-migration.adoc (hunk 7 above)
Comment (gnufied) [reply storage/container_storage_interface/persistent-storage-csi-migration.adoc:15]: I know - but I find it strange that we need to create "Additional Resources" link which refers to same page. Typically I would expect such links to take me somewhere else and provide more details.

Comment (lpettyjo) [reply storage/container_storage_interface/persistent-storage-csi-migration.adoc:17]: I can remove those xrefs if you prefer, but this is the only way to get around the fact that we cannot put xrefs in modules. The summary table refers to those subsections in text, but I cannot make those xrefs unfortunately.

File: modules/persistent-storage-csi-migration-sc.adoc [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+:_content-type: CONCEPT
+[id="persistent-storage-csi-migration-sc_{context}"]
+= Storage class
+
+== Storage class
+
+For new {product-title} 4.11 clusters, the default storage class is the CSI storage class.
+
+For clusters upgraded from {product-title} 4.10, or earlier, the CSI driver is used; however, the command output and other {product-title} indications show that the storage class of any upgraded cluster is the in-tree one (or whichever storage class that was defined as the default before the upgrade).
```
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-sc.adoc:13]: This has been corrected.

File: modules/persistent-storage-csi-migration-automatic-ga.adoc [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+* Azure Disk
+
+* OpenStack Cinder
+
+CSI migration for these volume types is considered generally available (GA), and requires no manual intervention.
+
+After upgrading from {product-title} version 4.10, and earlier, to 4.11, or later, the new default storage class is set to the CSI storage class. If you are installing a new cluster, the CSI storage class is the default, and all volumes provisioned using this storage class are CSI persistent volumes (PVs). The in-tree storage class remains, and might be necessary for certain features, such as volume expansion to work for existing in-tree PVs.
+
+However, for migrated volumes, the default storage class is not changed and remains the in-tree storage class, or whichever storage class was the default before the upgrade. While storage class referencing to the in-tree storage plug-in will continue working, we recommend that you switch the default storage class to the CSI storage class.
```
Comment (jeana-redhat) [inline modules/persistent-storage-csi-migration-automatic-ga.adoc:19]: I am a bit confused here - is there a difference between the upgrade scenario in the prior paragraph and the upgrade scenario in this paragraph?

File: modules/persistent-storage-csi-migration-enable.adoc [hunk 11]
```diff
@@ -4,7 +4,26 @@
 
 :_content-type: PROCEDURE
 [id="persistent-storage-csi-migration-enable_{context}"]
-= Enabling CSI automatic migration
+= Manually enabling CSI automatic migration
+
+If you want to test Container Storage Interface (CSI) migration in development or staging {product-title} clusters, in-tree to CSI migration must be manually enabled for the following in-tree volume types:
```
Comment (jeana-redhat) [inline modules/persistent-storage-csi-migration-enable.adoc:9]: Avoiding passive:
```suggestion
If you want to test Container Storage Interface (CSI) migration in development or staging {product-title} clusters, you must manually enable in-tree to CSI migration for the following in-tree volume types:
```

File: modules/persistent-storage-csi-migration-enable.adoc [hunk 12]
```diff
@@ -4,7 +4,26 @@
...
+
+* AWS Elastic Block Storage (EBS)
+
+* Google Compute Engine Persistent Disk (GCE-PD)
+
+* VMware vSphere Disk
+
+* Azure File
+
+CSI migration for these volume types is supported at the Technology Preview (TP) level.
```
Comment (jeana-redhat) [inline modules/persistent-storage-csi-migration-enable.adoc:19]: Usually stated this way. I am also wondering if perhaps you could just omit this line entirely since it's in the big red "READ ME" box?
```suggestion
CSI migration for these volume types is supported as a Technology Preview (TP) feature.
```

File: storage/container_storage_interface/persistent-storage-csi-migration.adoc [hunk 13]
```diff
@@ -6,13 +6,17 @@ include::_attributes/common-attributes.adoc[]
...
+=== Additional resources
+* xref:../../storage/container_storage_interface/persistent-storage-csi-migration.adoc#persistent-storage-csi-migration-automatic-ga_persistent-storage-csi-migration[Automatic migration of in-tree volumes to CSI]
+
+* xref:../../storage/container_storage_interface/persistent-storage-csi-migration.adoc#persistent-storage-csi-migration-enable_persistent-storage-csi-migration[Manually enabling CSI automatic migration]
+
+include::modules/persistent-storage-csi-migration-automatic-ga.adoc[leveloffset=+1]
+
 include::modules/persistent-storage-csi-migration-enable.adoc[leveloffset=+1]
 
 [role="_additional-resources"]
```
Comment (jeana-redhat) [inline storage/container_storage_interface/persistent-storage-csi-migration.adoc:22]: Git won't let me comment on the line below, but it should also be a `.Additional resources` style heading (it's tacking onto the preceding module, not an additional resource link for the entire assembly)

File: modules/persistent-storage-csi-migration-sc.adoc [hunk 14]
```diff
@@ -0,0 +1,17 @@
+// Module included in the following assemblies:
+//
+// * storage/container_storage_interface/persistent-storage-csi-migration.adoc
```
Comment (jeana-redhat) [inline modules/persistent-storage-csi-migration-sc.adoc:3]: I don't actually see this in the assembly.

**Review by jeana-redhat — COMMENTED** [review]
This is looking good - it's a difficult feature and supported status to cover! A few things but nothing too wild from me :)

**Comment by ropatil010** [comment]
Will it be valid to add in additional notes to include as IBM/Alibaba cloud platforms are default enabled as CSI migrated platforms because there are no intree provisioners for these platforms from version 4.10? 
cc: @gnufied @duanwei33  @lpettyjo

**Comment by ropatil010** [comment]
+ FYI, Default sc things attaching for Cinder/Azure disk, 
Azure cluster
Fresh cluster 4.10 or lesser => managed-premium (default) 
Fresh cluster 4.11 or later => managed-csi (default) 
Upgrade from 4.x to 4.11 => managed-premium (default)

Cinder cluster:
Fresh cluster 4.10 or lesser => standard (default)
Fresh cluster 4.11 or later => standard-csi (default)
Upgrade from 4.x to 4.11 => standard (default)

File: modules/persistent-storage-csi-migration-automatic-ga.adoc [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+{product-title} supports automatic and seamless migration for the following in-tree volume types to their Container Storage Interface (CSI) driver counterpart:
+
+* Azure Disk
+
+* OpenStack Cinder
+
+CSI migration for these volume types is considered generally available (GA), and requires no manual intervention.
+
+After upgrading from {product-title} version 4.10, and earlier, to 4.11, or later, the new default storage class is set to the CSI storage class. If you are installing a new cluster, the CSI storage class is the default, and all volumes provisioned using this storage class are CSI persistent volumes (PVs). The in-tree storage class remains, and might be necessary for certain features, such as volume expansion to work for existing in-tree PVs.
```
Comment (RomanBednar) [inline modules/persistent-storage-csi-migration-automatic-ga.adoc:17]:
```suggestion
If you are installing a new cluster, the CSI storage class is the default, and all volumes provisioned using this storage class are CSI persistent volumes (PVs).

After upgrading from {product-title} version 4.10, and earlier, to 4.11, or later, a new CSI storage class is added. This CSI storage class will be set as the default one only if there is no other default prior to upgrade. In a very unlikely case when there already is a storage class with a same name, the existing storage class will remain unchanged.
```

File: modules/persistent-storage-csi-migration-automatic-ga.adoc (hunk 10 above)
Comment (RomanBednar) [reply modules/persistent-storage-csi-migration-automatic-ga.adoc:19]: The upgrade scenario is the same. I think it's a bit confusing because first paragraph mentions upgrades and fresh installation and the second paragraph then refers again to upgrade scenario.

I'd reword this a bit to make a clear distinction between install and upgrade. And maybe add more details on how upgrade handles the storage classes. Because there's more scenarios that can happen while upgrading: users might already have some storage class that has a same name and would clash with our CSI one, or maybe they have some other default already set. So the fist sentence "After upgrading, the new default storage class is set..." is misleading because in fact it depends on pre-upgrade configuration.

Comment (RomanBednar) [inline modules/persistent-storage-csi-migration-automatic-ga.adoc:19]:
```suggestion
Any existing in-tree storage classes also remain unchanged, and might be necessary for certain features, such as volume expansion to work for existing in-tree PVs. For migrated volumes, the default storage class is not changed and remains the in-tree storage class, or whichever storage class was the default before the upgrade. While storage class referencing to the in-tree storage plug-in will continue working, we recommend that you switch the default storage class to the CSI storage class.
```

File: modules/persistent-storage-csi-migration-sc.adoc (hunk 14 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-sc.adoc:3]: Yup. I ended up not using it. I've deleted it now.

File: storage/container_storage_interface/persistent-storage-csi-migration.adoc [hunk 16]
```diff
@@ -6,13 +6,17 @@ include::_attributes/common-attributes.adoc[]
...
 
-{product-title} provides automatic migration for supported in-tree volume plug-ins to their equivalent Container Storage Interface (CSI) drivers.
-
-:FeatureName: CSI automatic migration
-include::snippets/technology-preview.adoc[leveloffset=+1]
+In-tree storage drivers that are traditionally shipped with {product-title} are being deprecated and replaced by their equivalent Container Storage Interface (CSI) drivers. {product-title} provides automatic migration for certain supported in-tree volume plug-ins to their equivalent CSI drivers.
 
 include::modules/persistent-storage-csi-migration-overview.adoc[leveloffset=+1]
 
+=== Additional resources
```
[the remark this answers is not in the document]
Comment (lpettyjo) [reply storage/container_storage_interface/persistent-storage-csi-migration.adoc:13]: Done

File: modules/persistent-storage-csi-migration-overview.adoc [hunk 17]
```diff
@@ -6,27 +6,30 @@
 [id="persistent-storage-csi-migration-overview_{context}"]
 = Overview
 
-With the automatic migration feature enabled, volumes provisioned using in-tree storage plug-ins supported by this feature are migrated to their counterpart CSI drivers.
+Volumes provisioned by using in-tree storage plug-ins, and that are supported by this feature, are migrated to their counterpart Container Storage Interface (CSI) drivers. This process does not perform any data migration; {product-title} only translates the persistent volume object in memory. As a result, the translated persistent volume object is not stored on disk, nor is its contents changed.
```
[the remark this answers is not in the document]
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-overview.adoc:9]: Done

File: modules/persistent-storage-csi-migration-enable.adoc (hunk 11 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-enable.adoc:9]: Done

File: modules/persistent-storage-csi-migration-enable.adoc (hunk 12 above)
Comment (lpettyjo) [reply modules/persistent-storage-csi-migration-enable.adoc:19]: Removed sentence as added this statement to TP boilerplate note.

**Comment by lpettyjo** [comment]
> * FYI, Default sc things attaching for Cinder/Azure disk,
>   Azure cluster
>   Fresh cluster 4.10 or lesser => managed-premium (default)
>   Fresh cluster 4.11 or later => managed-csi (default)
>   Upgrade from 4.x to 4.11 => managed-premium (default)
> 
> Cinder cluster: Fresh cluster 4.10 or lesser => standard (default) Fresh cluster 4.11 or later => standard-csi (default) Upgrade from 4.x to 4.11 => standard (default)

@ropatil010 I believe that this content is covered in the last two paragraphs of the "Automatic migration of in-tree volumes to CSI" section. Please let me know if this is not the case. Thanks for your review!

**Comment by gcharot** [comment]
looks good to me, thanks!

**Comment by bertinatto** [comment]
@lpettyjo LGTM!

**Comment by dobsonj** [comment]
LGTM

**Comment by ropatil010** [comment]
LGTM
[Optional] Regarding #2 point, on Migrated platforms, if we again apply the feature gate patch, it takes 5min to bring the cluster back in healthy condition. 
Thanks @lpettyjo
